storage/pebble/procedure/cluster.go

```diff
@@ -113,14 +116,14 @@ func FinalizeClusterBlock(blockID flow.Identifier) func(*badger.Txn) error {
 
 		// retrieve the current finalized state boundary
 		var boundary uint64
-		err = operation.RetrieveClusterFinalizedHeight(chainID, &boundary)(tx)
+		err = operation.RetrieveClusterFinalizedHeight(chainID, &boundary)(r)
 		if err != nil {
 			return fmt.Errorf("could not retrieve boundary: %w", err)
 		}
 
 		// retrieve the ID of the boundary head
 		var headID flow.Identifier
-		err = operation.LookupClusterBlockHeight(chainID, boundary, &headID)(tx)
+		err = operation.LookupClusterBlockHeight(chainID, boundary, &headID)(r)
 		if err != nil {
 			return fmt.Errorf("could not retrieve head: %w", err)
 		}
```

Comment:
## ⚠️ 

[lines 124 - 134](https://github.com/onflow/flow-go/blob/51969ecf92ea9775622890d6be9be223f6f7a593/storage/pebble/procedure/cluster.go#L124-L134) are always subject to stale reads with pebble, and therefore this implementation insufficiently guarantees state consistency. 

It is analogous problem to the one described in [my previous comment](https://github.com/onflow/flow-go/pull/6197#discussion_r1716273405). I think we need a refactoring here, where we move the logic to inserting blocks and indexing children to _one specific component_ that has the _sole authority_ to maintain these indices in a consistent manner.